### PR TITLE
fix(pacquet): ignore a broken wanted lockfile on regular installs

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Root-cause fix for the "bd install destroys the bit workspace's node_modules" investigation.

What actually happened in that incident, in order:

1. bit's `.modules.yaml` had been written by the TypeScript engine (different `storeDir`/`virtualStoreDir` era), so the next pacquet-engine install legitimately detected layout drift and purged `node_modules` — the same decision `validateModules` makes upstream, so no parity issue there.
2. The install then tried to load `pnpm-lock.yaml`, and the parse failed (on the older engine, the since-fixed node-budget cap; but any parse failure reproduces it). pacquet treated the broken wanted lockfile as **fatal for every install flavor**, so the run aborted *after* the purge — leaving the workspace with no packages and no way to reinstall.

Step 2 is a parity bug and is what this PR fixes: upstream's `readLockfiles` only fails on a broken wanted lockfile under `--frozen-lockfile`; a regular install warns `Ignoring broken lockfile at <dir>: …`, resolves from the manifests, and rewrites the lockfile. pacquet now does the same (the *current* lockfile arm already behaved this way). With this fix, that incident degrades to a warning plus one full re-resolution instead of a destroyed workspace.

Also ported the missing TS test for the ignore path (`a lockfile with duplicate keys is fixed`, `deps-installer/test/lockfile.ts:1288`) as `a_broken_wanted_lockfile_is_ignored_and_regenerated`, next to the already-ported frozen variant, and verified it discriminates (fails with the fix reverted). Updated `plans/TEST_PORTING.md`.

Not a pnpm bug, for the record: the *other* wreckage mode (healthy state → legitimate purge → engine completes → bd dies lazily loading an aspect whose compiled `dist` the purge removed) reproduces with the purge semantics both stacks share — it is inherent to bd running from the workspace it is reinstalling, and needs a bit-side answer (e.g. preloading aspects before install).

## Squash Commit Body

```text
A pnpm-lock.yaml that fails to parse aborted every install flavor,
matching upstream only for --frozen-lockfile. A regular install now
warns ("Ignoring broken lockfile at ..."), resolves from the
manifests, and rewrites the file - the lockfile is regenerable state,
and failing left no way to install at all. Found via Bit: a broken
lockfile aborted the install after node_modules had already been
purged for layout drift, leaving the workspace with neither packages
nor a usable lockfile.

Ports deps-installer/test/lockfile.ts:1288 ("a lockfile with duplicate
keys is fixed") as the regression test.
```

## Checklist

- [x] The change is implemented in both the TypeScript CLI and the Rust
  `pnpm/` port, or the description notes what still needs porting.
  (Rust-only: this aligns pacquet with existing TypeScript behavior.)
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published
  package.
- [x] Added or updated tests.

---
Written by an agent (Claude Code, claude-fable-5).

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/13511"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1788034671&installation_model_id=426870&pr_number=13511&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F13511&signature=333c084a19f1e9061ed498374e0e5781bd5baddec50030058611651709fb051e"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Regular `pnpm install` now succeeds when `pnpm-lock.yaml` exists but is corrupted/unparseable, emitting an “Ignoring broken lockfile” warning and regenerating the lockfile from project manifests.
  * `pnpm install --frozen-lockfile` still fails when the lockfile is invalid.
* **Documentation**
  * Updated test/porting documentation and added/ported coverage for broken wanted lockfile scenarios, including duplicate keys.

### CODE REVIEW

File: pnpm/crates/package-manager/src/install.rs [hunk 1]
```diff
@@ -1114,7 +1114,27 @@ where
...
+                    message: format!(
+                        "Ignoring broken lockfile at {}: {error}",
+                        workspace_root.display(),
+                    ),
+                    prefix: prefix.clone(),
+                }));
+                None
+            }
+            Err(error) => return Err(InstallError::LoadWantedLockfile(error)),
+        };
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/install.rs]: Verified against the actual control flow and empirically — the shortcut cannot mask the broken lockfile:

1. The optimistic path probes `wanted_lockfile_modified` (mtime vs the state's `last_validated_timestamp`) before it may conclude up-to-date, so a lockfile corrupted after the last install always defeats the pure-mtime exit.
2. When the fast path does need lockfile contents, `modified_manifests_match_lockfile` maps a parse failure to `Decision::Skipped` ("the wanted lockfile cannot be read or parsed") — the run falls through to the full pipeline, which is where the new recovery arm warns and regenerates.

End-to-end check with the **default** setting (no `optimisticRepeatInstall: false`): install → append a duplicate key to `pnpm-lock.yaml` → `pnpm install` prints `[WARN] Ignoring broken lockfile at …: duplicate mapping key`, succeeds, and rewrites a valid lockfile — no "Already up to date".

The test disables the shortcut for the same reason the neighboring ported test (`a_broken_current_lockfile_is_ignored_with_a_warning`) does: determinism. The corruption is written within the same clock tick as the install on coarse-mtime filesystems, so the mtime probe's verdict — and therefore *which* code path detects the breakage — would be timing-dependent. The test pins the recovery arm itself; the mtime gate is the fast path's existing, separately-covered contract.

---
Written by an agent (Claude Code, claude-fable-5).

**Comment by zkochan** [comment]
Addressed the bot findings:

1. **Changeset missing `pnpm` entry (Qodo rule 713436)** — intentionally omitted. Repo policy (AGENTS.md, "Changesets for the Rust products") is the opposite of the rule for Rust-only changes: *"A changeset whose implementation is Rust-only and targets `pacquet` must omit `pnpm`"* — in changesets, `pnpm` means the TypeScript CLI package, which this PR does not touch (it already has this behavior; the fix brings pacquet to parity).
2. **Raw YAML append in the test** — fixed in 8d9c35cd7c: the test now uses the existing `append_workspace_yaml_key` helper, which is newline-safe and guards against duplicate keys.
3. **Unsanitized parser diagnostic in the warning** — keeping as is: the message shape mirrors both the TypeScript CLI (`Ignoring broken lockfile at ${dir}: ${err.message}` with the raw parser message) and pacquet's existing current-lockfile arm; diverging here would break log parity, and the same sanitization request has been rejected in prior reviews.

CodeRabbit's optimistic-fast-path concern is answered on its thread: the shortcut probes the wanted lockfile's mtime before concluding up-to-date and maps parse failures to a skip, verified end-to-end with the default setting.

---
Written by an agent (Claude Code, claude-fable-5).
